Do not extract ISBN-10s from hyphenated ISBN-13s

mudge opened the pull request:
The improved support for hyphenated ISBNs has accidentally started extracting ISBN-10s from ISBN-13s when they are substrings of one another (typically the check digit changes but there are some ISBNs with check digits that do not change when converting).

As this means we'll start returning duplicate ISBNs for any ISBN that has this property, explicitly prevent extracting ISBN-10s from a longer hyphenated string of digits.

mudge committed:
Do not extract ISBN-10s from hyphenated ISBN-13s

The improved support for hyphenated ISBNs has accidentally started
extracting ISBN-10s from ISBN-13s when they are substrings of one
another (typically the check digit changes but there are some ISBNs
with check digits that do not change when converting).

As this means we'll start returning duplicate ISBNs for any ISBN that
has this property, explicitly prevent extracting ISBN-10s from a longer
hyphenated string of digits.

jbilbo commented on lib/identifiers/isbn.rb:
```diff
@@ -13,6 +13,7 @@ class ISBN
     /x
     ISBN_10_REGEXP = /
       \b
+      (?<!\p{Pd})       # Do not accidentally match a hyphenated ISBN-13
```
Tricky! Is there something we can do when the ISBN is separated by spaces?

mudge replied:
Good question: I could make this stricter (don't match an ISBN-10 if it is preceded by a number followed by a space or a dash) but negative lookbehinds are quite restrictive in what you can do.

I'll experiment.

mudge committed:
Do not extract ISBN-10s from spaced ISBN-13s

As well as preventing valid ISBN-10s being extracted from ISBN-13s
separated by dashes, prevent ISBN-10s being extracted from ISBN-13s
separated by spaces.

In order to achieve this, we have to be more explicit about the format
of ISBN-13s by looking for the ISBN (GS1) Bookland prefix.

jbilbo reviewed (approved):
✨

jbilbo commented on lib/identifiers/isbn.rb:
```diff
@@ -12,6 +12,10 @@ class ISBN
       \b
     /x
     ISBN_10_REGEXP = /
+      (?<!              # Do not accidentally match a hyphenated ISBN-13
```
too late, I know. But could you update this comment since now also takes into account a spaced ISBN-13
